### ISSUE TITLE
nn.CrossEntropy/nn.NLLLoss : Request for option to specify invalid ignore_index for perf. optimization

### ISSUE THREAD

**Issue by jayachandranb-ai** [pytorch/pytorch#89907]
### 🚀 The feature, motivation and pitch

nn.CrossEentropy and nn.NLLLoss functions have the argument "ignore_index" to specify the target index to be ignored during loss calculation
Eg: Pls ref:
https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html 

The default value is -100 and as per the definition at the above link, any value (even outside the usual class range) can be specified as ignore_index value.
And there is no option to specify that ingore_index value is not valid.( i.e, All elements of target are to be considered for loss calculation)
In fact some models seem to intentionally  set targets to be ignored as -100 since default ignore_index is -100.
(https://github.com/huggingface/transformers/blob/main/src/transformers/models/layoutxlm/tokenization_layoutxlm.py)

Looking for whether the target tensor contains any element with ignore_index value can have computational overhead especially with large target tensors.
But there is no target to be ignored in the case of many models.
So the current API results in unnecessary overhead in many cases.

Can PyTorch provide an option to indicate the backends whether the inore_index argument is invalid
(so that sifting though the target tensor for ignore_index can be avoided)
Say, with an API update with a "specific" ingore_index value or an additional argument?

This can improve the training times if going forward backends and training scripts can adopt this change.

When the default value was set as -100, was it intended to be used as as an invalid_ignore index?
Can models use -100 as invalid ignore_index? If yes, may be just a document update is enough from Pytorch side.
And hopefully existing models can make corresponding change when upgrading for a Pytorch version with this specification.

**Comment by wconstab** [comment]
This seems reasonable to me, but tagging @gchanan to weigh in or pull in other POCs.

**Comment by ngimel** [comment]
Do you have any evidence for your claim that handling `ignore_index` actually affects performance? Typically loss computation is a very small portion of e2e computation, and furthermore, in the implementation the target values are read (they have to be read regardless), compared with `ignore_index` (a very fast operation, negligible effect on performance) and kernel continues. There are no additional kernel launches, no additional memory reads, no branch divergence.

**Comment by jayachandranb-ai** [comment]
Thanks @ngimel, @wconstab ..

It is the comparison of ignore index with target values(and reading target for this purpose) that causes perf overhead.
Especially, in the "mean" mode of CrossEntropy/NLLLoss calculation, the scale factor
for mean needs to be found based on the number of "ignore_index valued labels" present in the target tensor.
In parallel implementations of loss kernel, since data is distributed among multiple processing elements,
each proc. element needs to know this scale factor. This increases the number of memory reads and
can lead to bandwidth pressures for large tensors.
While loss itself is a small portion of overall e2e calculation as you mentioned, it still helps to avoid this extra overhead.
Hence it is better to give backends an explicit indication if ignore_index has to be necessarily considered.
Backend implementations that can/need to make use if this extra info can benefit from it.

**Comment by ngimel** [comment]
As I said, you've presented no evidence that it would result in meaningful performance improvement.
